Accessibility: Move focus back to the modal dialog to More Menu when it was used to get there

gziolo opened the pull request:
## Description

Related: #15321 (it's probably the fix).

> When the "Keyboard Shortcuts" and "Options" dialogs are dismissed,
the keyboard focus is reset to the top of the page. This may be
confusing or inconvenient for users, who would expect the focus to
return to the last element they were focused on.

This PR changes the behavior of More Menu in terms of maintaining focus. When a dialog is opened from one of the menu items, then we no longer close More Menu. Instead, we keep it opened and move focus back to the More Menu when the dialog is closed.

### Known issue

~When I mouse-click outside of the modal dialog then More Menu remains opened. This is because Modal dialog prevents all events to be propagated below the overlay layer. We need to find a fix for it, @youknowriad any hints?~ - this is resolved now

## How has this been tested?

1. Open More Menu.
2. Click on the Block Manager menu item (or anything else).
3. Close the dialog by clicking on the close button.
4. Focus should return to the item in More Menu.
5. Repeat steps 2-4 but this time click outside of the modal dialog instead..
6. Repeat using keyboard-only navigation (space to select an item and `esc` to close the dialog.

See also the screencast.

## Screenshots <!-- if applicable -->
![focus-more-menu-dialog](https://user-images.githubusercontent.com/699132/62704049-ccbdfe00-b9ea-11e9-8e01-9d60677ead04.gif)


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

gziolo committed:
Accessibility: Move focus back to the modal dialog to More Menu when it was used to get there

gziolo commented on packages/components/src/dropdown/index.js:
```diff
@@ -49,10 +49,14 @@ class Dropdown extends Component {
-		if ( ! this.containerRef.current.contains( document.activeElement ) ) {
+		if (
+			! this.containerRef.current.contains( document.activeElement ) &&
+			! document.activeElement.closest( '[role="dialog"]' )
```
I have no idea whether we can assume that focus can be moved to the element with `[role="dialog"]` in all cases. Thoughts?

talldan replied:
It definitely feels like there needs to be a more general way to track that a modal or popover was opened from the context of another modal or popover, and exclude it from the focus outside check.

There's a similar issue here with the inline link popover, where it needs to check whether focus is in the autocomplete list to avoid closing:
https://github.com/WordPress/gutenberg/blob/fb910e55e021b66be8332fb979536d885e87ab95/packages/format-library/src/link/inline.js#L168-L179

Having said that, I can't think of a case where this logic fails, so I think it'd be pragmatic to ship this, and then consider what a more general solution might be.

gziolo commented:
> When I mouse-click outside of the modal dialog then More Menu remains opened. This is because Modal dialog prevents all events to be propagated below the overlay layer. We need to find a fix for it, @youknowriad any hints?

This is exactly how More Menu stays wrongly opened:

![more-menu-open](https://user-images.githubusercontent.com/699132/62705316-71d9d600-b9ed-11e9-88a3-22d264c0f549.gif)

afercia commented:
> @afercia - do we have an open issue for this or did we talk about it only in comments on other PRs/issues?

@gziolo Besides mentions in other PRs/issues, I think https://github.com/WordPress/gutenberg/issues/15321 is very specific to this problem.

gziolo commented:
> @gziolo Besides mentions in other PRs/issues, I think #15321 is very specific to this problem.

Thank you, I need to figure out how you search for those issues, there are too many of them :) Yes, I think this PR solves this exact problem. There is a similar issue for popovers like a calendar as well.

gziolo commented:
> > When I mouse-click outside of the modal dialog then More Menu remains opened. This is because Modal dialog prevents all events to be propagated below the overlay layer. We need to find a fix for it, @youknowriad any hints?
> 
> This is exactly how More Menu stays wrongly opened:
> 
> ![more-menu-open](https://user-images.githubusercontent.com/699132/62705316-71d9d600-b9ed-11e9-88a3-22d264c0f549.gif)

Once #17297 gets merged, this issues will automatically get resolved 🎉

gziolo committed:
Refactor tests to make them work with refined behavior of the modal component with the dropdown menu

gziolo commented:
I rebased this branch with `master` and refactored some test utils and tests which use them to make everything work with the updated behavior of the dropdown menu and modal dialogs.

This PR is ready for the final review and testing.

talldan reviewed (approved):
Works really nicely in testing 🎉
